[For 10.4] Add share indicator in webUI

phil-davis opened the pull request:
## Description
This is a rebased and squashed version of #36254 (which has many commits)

Filelist shows share indicators on files and folders, that reside inside a shared folder / shared folders.
The sidebar share-tab reveals a detailed view of the share-tree including share-recipients and the respective folders the life in.

![2019-11-12-21-23-13](https://user-images.githubusercontent.com/12717530/68707852-d54a4580-0592-11ea-9585-819370243361.gif)

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/enterprise/issues/3453 (Discussion)
https://github.com/owncloud/enterprise/issues/3556 (Vis. Concept)

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Manual testing in
- Google Chome
- Mozilla Firefox
- MSIE 11

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](../changelog/README.md)

lordelix committed:
Add dir/path sharing info

kiranparajuli589 committed:
Add acceptance tests for the share indicators on the webUI

phil-davis commented:
ToDo: I need to review the tests in the 2nd commit.

individual-it commented on tests/acceptance/features/lib/FilesPage.php:
```diff
@@ -399,4 +400,17 @@ public function browseToFolder($folderName) {
+	 * @return bool
+	 */
+	public function isShareIndicatorPresent($fileName) {
+		$resourceMarkedSharedXpath = \sprintf($this->resourceSharedIndicatorXpath, $fileName);
```
use `findFileRowByName`

individual-it replied:
fixed in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/lib/SharedByLinkPage.php:
```diff
@@ -104,6 +114,25 @@ public function __construct(
+	 *
+	 * @throws ElementNotFoundException
+	 */
+	public function findFileRowByNameAndPath($name, $path, Session $session) {
```
unneeded

individual-it replied:
deleted in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature:
```diff
@@ -677,3 +677,192 @@ Feature: Sharing files and folders with internal users
+      | user1    |
+      | user2    |
+    And user "user2" has logged in using the webUI
+    When the user shares folder "simple-folder" with user "User One" using the webUI
```
sharing should be done by API

individual-it replied:
fixed in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature:
```diff
@@ -677,3 +677,192 @@ Feature: Sharing files and folders with internal users
       just letting you know that user0 shared simple-folder with you.
       """
+
+  Scenario: Create share with internal users and view those shared resources in files page and shared-with-others page
```
views don't make a difference, because we have to go into a folder

individual-it replied:
deleted unneeded tests in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature:
```diff
@@ -677,3 +677,192 @@ Feature: Sharing files and folders with internal users
+      | testavatar.png                        |
+      | testapp.zip                           |
+
+  Scenario: user creates a public link for a folder and views the shared folder from files page, shared-with-others page
```
public link has to be put in other feature file

individual-it commented on tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature:
```diff
@@ -677,3 +677,192 @@ Feature: Sharing files and folders with internal users
+      | simple-folder                   |
+
+  Scenario: user receives a share from another user and views that received share in files page and shared-wih-you page when auto-accept-share is enabled
+    Given the setting "Automatically accept new incoming local user shares" in the section "Sharing" has been enabled
```
that is default

individual-it replied:
no need to test in other views in the first place, deleted in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature:
```diff
@@ -677,3 +677,192 @@ Feature: Sharing files and folders with internal users
+      | testavatar.png                        |
+      | testapp.zip                           |
+
+  Scenario: Create share for a folder within a group and view the shared folder in files page and shared-with-others page
```
group shares are is separate feature file

individual-it replied:
fixed in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it committed:
simplify tests

individual-it committed:
forgotten file upload

individual-it committed:
revert changes to existing test

individual-it commented on tests/acceptance/features/bootstrap/WebUIFilesContext.php:
```diff
@@ -1418,7 +1434,6 @@ public function checkIfFileFolderIsListedOnTheWebUI(
 			case "shared-by-link page":
 				$this->theUserBrowsesToThePage($this->sharedByLinkPage);
 				break;
-			case "shared with others page":
```
fixed in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/bootstrap/WebUISharingContext.php:
```diff
@@ -1601,4 +1601,68 @@ public function tearDownScenario() {
+	 * @return void
+	 * @throws Exception
+	 */
+	public function folderShouldHaveShareIndicatorOnTheWebUI($fileOrFolder, $filename) {
```
fixed in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/bootstrap/WebUISharingContext.php:
```diff
@@ -1601,4 +1601,68 @@ public function tearDownScenario() {
 	}
+
+	/**
+	 * @Then /^(file|folder) "([^"]*)" should have share indicator on the webUI$/
```
deleted the complete function in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/bootstrap/WebUISharingContext.php:
```diff
@@ -1601,4 +1601,68 @@ public function tearDownScenario() {
+		$isMarked = $this->filesPage->isShareIndicatorPresent(
+			$filename
+		);
+		Assert::assertEquals(true, $isMarked, "Expected: " . $fileOrFolder . " to be marked as shared but found: " . \strval($isMarked));
```
fixed in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented on tests/acceptance/features/bootstrap/WebUISharingContext.php:
```diff
@@ -1601,4 +1601,68 @@ public function tearDownScenario() {
+	/**
+	 * @Then the following resources should not have share indicators on the webUI
+	 *
+	 * @param TableNode $resourceTable table headings: must be: |name|
```
fixed in 448552d664b5db9ac8837b67495c2f5be13bd0ec

individual-it commented:
ToDo left for tests:
- [ ] test if uploading file in a shared folder shows the indicator immediately
- [ ] test if creating a folder in a shared folder shows the indicator immediately
- [ ] test with public links
- [ ] test with federation shares

individual-it committed:
Merge branch 'master' into feature/path-share-info

individual-it committed:
add @skipOnOcV10.3 tags on new tests

individual-it commented:
- [ ]  details view does not indicate if the share receiver is a group or user as usual in the sharing dialogues
![image](https://user-images.githubusercontent.com/2425577/70389242-d715f600-19e4-11ea-8648-ce112ff3b2d4.png)

individual-it commented:
- [ ]  reshares are not shown in the details for the original sharer like in the share list
![image](https://user-images.githubusercontent.com/2425577/70389675-b2bd1800-19ea-11ea-8bca-1741b931a952.png)
vs
![image](https://user-images.githubusercontent.com/2425577/70389679-c0729d80-19ea-11ea-9dd8-950e1fc77fa1.png)

both screenshots are of the original sharer, but uu1 has reshared the folder to uu2

individual-it committed:
first part of testing the details tab

kiranparajuli589 committed:
Add acceptance test for uplading files and creating folder inside a shared folder to check share-indicator

phil-davis commented on tests/acceptance/features/webUISharingInternalGroups/shareWithGroups.feature:
```diff
@@ -316,6 +316,7 @@ Feature: Sharing files and folders with internal groups
     And the user opens the share dialog for file "lorem.txt"
     Then the group "grp1" should not be in share with group list
 
+<<<<<<< HEAD
```
leftover from conflict

kiranparajuli589 replied:
fixed!
:rocket:

phil-davis commented on tests/acceptance/features/webUISharingInternalUsers/shareWithUsers.feature:
```diff
@@ -746,6 +746,7 @@ Feature: Sharing files and folders with internal users
     And user "user1" has uploaded file "filesForUpload/lorem.txt" to "/simple-folder/lorem.txt"
     And user "user1" has shared folder "simple-folder" with user "user2"
     And user "user1" has logged in using the webUI
+<<<<<<< HEAD
```
leftover from conflict

kiranparajuli589 replied:
:rocket:

phil-davis reviewed (commented):
leftovers from conflicts must be cleaned up

kiranparajuli589 committed:
fixes

kiranparajuli589 committed:
Acceptance test for share indicator inside folder shared using public link

kiranparajuli589 committed:
Scenario added for share-indicator inside a shared folder using public link

kiranparajuli589 committed:
minor fixes

kiranparajuli589 committed:
Acceptance test for share indicator inside folder shared using federated sharing

lordelix committed:
Display reshares as well

kiranparajuli589 committed:
Acceptance test for sharing details in different and multiple share-types

kiranparajuli589 committed:
fixups

lordelix committed:
Add additional for remote shares
ellipsis overflowing text

felixheidecke commented:
@individual-it I addressed the two issues above in this PR:
https://github.com/owncloud/core/pull/36549

individual-it committed:
Merge branch 'feature/path-share-info' into feature/path-share-info-fix-group-reshare

individual-it committed:
adjust acceptance tests for fixed issues

individual-it committed:
Merge pull request #36549 from owncloud/feature/path-share-info-fix-group-reshare

Feature/path share info fix group reshare

jasson99 commented:
[](file:///home/jasminebaral/Pictures/Screenshot%20from%202019-12-10%2011-55-06.png)

jasson99 commented:
Public links created without name through webUI or using  sharing API are listed without any name on the webUI as shown below.


![Screenshot from 2019-12-10 11-55-06](https://user-images.githubusercontent.com/40829116/70500818-89180400-1b44-11ea-81f1-9d1fcc5fa2df.png)

jasson99 committed:
Acceptance test for share details for federation sharing

lordelix committed:
Show link token if name is empty

felixheidecke commented:
@jasson99 @individual-it I addressed the issue with the missing name for public links here:
https://github.com/owncloud/core/pull/36559

kiranparajuli589 committed:
Acceptance test added for sharing details when multiple public links with strange link names are created

kiranparajuli589 committed:
Added acceptance test for sharing detail of public link shares created with empty name

individual-it committed:
Merge pull request #36561 from owncloud/share-detail-public-link

Acceptance test added for sharing details when multiple public links …

individual-it committed:
Merge branch 'feature/path-share-info' into sharing-details-in-federation-sharing

individual-it committed:
Merge pull request #36562 from owncloud/sharing-details-in-federation-sharing

Acceptance test for share details for federation sharing

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/core?pullRequest=36534) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/core?pullRequest=36534) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: felixheidecke<br/>:white_check_mark: kiranparajuli589<br/>:white_check_mark: individual-it<br/>:x: jasson99<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/core?pullRequest=36534) it.</sub>

individual-it commented on tests/acceptance/features/webUISharingPublic/shareByPublicLink.feature:
```diff
@@ -662,4 +662,74 @@ Feature: Share by public link
+    Then the following resources should have share indicators on the webUI
+      | sub-folder |
+
+  Scenario: sharing details of items inside a shared folder
```
Need to be skipped on 10.3

individual-it replied:
done in https://github.com/owncloud/core/pull/36534/commits/7bc6d3515fc5dc8e6f2dd2aa0e690781f8a48177

kiranparajuli589 replied:
this is fixed in ##36559

individual-it committed:
fix step sentence was adjusted in 51d37693744d52aa00d56b3e75ce6e700b3fd7f3

individual-it committed:
skip one more test on 10.3

kiranparajuli589 committed:
fixing conflicts

individual-it committed:
Merge pull request #36559 from owncloud/feature/path-share-info-link-name-alternative

Show link token if name is empty

individual-it commented:
@micbar @felixheidecke @phil-davis I think acceptance-test wise we are ready here, taking off the QA-team label. Please review if the tests are what you would expect from the feature

phil-davis commented:
See PR #36572 - I have squashed and rebased this long list of commits. That PR has the same code on top of current master. Please do final reviews there.
(A simple squash-rebase here gave me various conflicts during the rebase process and I wanted to make sure to not lose anything that is already on this branch)

I will close here after CI finishes.
